Add pos_weight argument to nn.BCEWithLogitsLoss (#5660)

velikodniy opened the pull request:
- Add an option to control precision/recall in imbalanced datasets
- Add tests (but new_criterion_tests)

Multiplier `1 + (pos_weight - 1) * target` is the only significant difference.

Notes:

- I didn't implement pos_weight for `F.binary_cross_entropy`/`nn.BCELoss`. (It uses implementation from torch._C.) I can add a straightforward and numerically unstable implementation to this function but I'm not sure if it is really needed.
- I've added `pos_weight` as *last* argument to prevent errors in the code that doesn't use names for keyword arguments. But it looks quite ugly.

P.S. I proposed these changes in #5660

velikodniy committed:
Add pos_weight argument to nn.BCEWithLogitsLoss and F.binary_cross_entropy_with_logits (#5660)
- Add an option to control precision/recall in imbalanced datasets
- Add tests (but new_criterion_tests)

ezyang commented on torch/nn/functional.py:
```diff
@@ -1507,6 +1507,8 @@ def binary_cross_entropy_with_logits(input, target, weight=None, size_average=Tr
         target: Tensor of the same shape as input
         weight (Tensor, optional): a manual rescaling weight
                 if provided it's repeated to match input tensor shape
+        pos_weight (Tensor, optional): a weight of positive examples.
```
Isn't this the wrong place to put the kwarg? It should go at the end.

ezyang commented on torch/nn/modules/loss.py:
```diff
@@ -458,10 +458,26 @@ class BCEWithLogitsLoss(_Loss):
         weight (Tensor, optional): a manual rescaling weight given to the loss
             of each batch element. If given, has to be a Tensor of size
             "nbatch".
+        pos_weight (Tensor, optional): a weight of positive examples.
```
Ditto here

ezyang reviewed (approved):
Accept assuming the doc issues are fixed

velikodniy committed:
Move pos_weight to the end of args list in the documentation.

`pos_weight` was moved to the end because it is the last argument in both
`nn.BCEWithLogitsLoss` and `binary_cross_entropy_with_logits`

apaszke commented on torch/nn/functional.py:
```diff
@@ -1515,6 +1515,8 @@ def binary_cross_entropy_with_logits(input, target, weight=None, size_average=Tr
                 is ``False``, returns a loss per input/target element instead and ignores
                 :attr:`size_average`. Default: ``True``
+        pos_weight (Tensor, optional): a weight of positive examples.
+                Must be a vector with length equal to the number of classes.
```
Isn't the number of classes always 2 in this case? The description is pretty confusing

velikodniy replied:
Of course, you're right. BCE is binary by definition. I meant multilabel classification.

What about this?

> Must be a vector with length equal to the number of labels where
> each label is corresponding to a binary class.

apaszke replied:
Sorry, I still don't think I understand. This loss wants an input and target of the same shape, but the target should be a binary tensor. Now, how should `pos_weight` look like? Is it just a 2 element tensor?

apaszke replied:
Also, I'm wondering if we really want to add this. You can easily weight samples according to even more complicated schemes if you only use `reduce=False`. In particular, I think this one is equivalent to the code below, which isn't too complicated.
```python
(F.binary_cross_entropy_with_logits(..., reduce=False) * class_weights).mean()
```
Here `class_weights` allows you to weight both things, if you want to get the `pos_weight` behavior then you do
```python
class_weights = torch.ones_like(input).mask_fill_(target == 1, pos_weight_value)
```

velikodniy replied:
`class_weights` are the weights of labels. `pos_weight` acts *inside* BCE formula for every particular label.

**The usual BCE**:

`target * -log(sigmoid(input)) + (1 - target) * -log(1 - sigmoid(input))`

For multilabel classification we can use vectors for `target` and `input`.

**The proposed BCE**:

`pos_weight * target * -log(sigmoid(input)) + (1 - target) * -log(1 - sigmoid(input))`

`pos_weight` is multiplied by only the "positive" part. It allows us to "increase" the fraction of positive examples in imbalanced datasets, so we can trade off precision and recall. I don't know any simple way to do it *outside* BCE.

TF already have [a similar function](https://www.tensorflow.org/api_docs/python/tf/nn/weighted_cross_entropy_with_logits).

For example. If we have 7 labels and the batch size is equal to 4. Then:
- input — 4×7
- target — 4×7
- pos_weight — 1×7 (or just 7, thanks to broadcasting) — a vector of weights of positive part for every label.

apaszke commented on torch/nn/modules/loss.py:
```diff
@@ -484,20 +500,17 @@ class BCEWithLogitsLoss(_Loss):
+    def __init__(self, weight=None, size_average=True, reduce=True, pos_weight=None):
         super(BCEWithLogitsLoss, self).__init__(size_average, reduce)
         self.register_buffer('weight', weight)
+        self.register_buffer('pos_weight', pos_weight)
```
This will break backward compatibility. We need to implement `__setattr__`, and some extra care around state dicts might be necessary too.

velikodniy replied:
Is it needed for loading state dict? I'm not sure if I properly understand the problem.

SsnL replied:
@apaszke I don't think that this will break BC because we don't save & load None buffers anyways.

ezyang commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please
